##### Copyright 2020 The TensorFlow Authors.

In [1]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Multi-task recommenders

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://www.tensorflow.org/recommenders/examples/multitask"><img src="https://www.tensorflow.org/images/tf_logo_32px.png" />View on TensorFlow.org</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/tensorflow/recommenders/blob/main/docs/examples/multitask.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/tensorflow/recommenders/blob/main/docs/examples/multitask.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View source on GitHub</a>
  </td>
  <td>
    <a href="https://storage.googleapis.com/tensorflow_docs/recommenders/docs/examples/multitask.ipynb"><img src="https://www.tensorflow.org/images/download_logo_32px.png" />Download notebook</a>
  </td>
</table>

In the [basic retrieval tutorial](basic_retrieval) we built a retrieval system using movie watches as positive interaction signals.

In many applications, however, there are multiple rich sources of feedback to draw upon. For example, an e-commerce site may record user visits to product pages (abundant, but relatively low signal), image clicks, adding to cart, and, finally, purchases. It may even record post-purchase signals such as reviews and returns.

Integrating all these different forms of feedback is critical to building systems that users love to use, and that do not optimize for any one metric at the expense of overall performance.

In addition, building a joint model for multiple tasks may produce better results than building a number of task-specific models. This is especially true where some data is abundant (for example, clicks), and some data is sparse (purchases, returns, manual reviews). In those scenarios, a joint model may be able to use representations learned from the abundant task to improve its predictions on the sparse task via a phenomenon known as [transfer learning](https://en.wikipedia.org/wiki/Transfer_learning). For example, [this paper](https://openreview.net/pdf?id=SJxPVcSonN) shows that a model predicting explicit user ratings from sparse user surveys can be substantially improved by adding an auxiliary task that uses abundant click log data.

In this tutorial, we are going to build a multi-objective recommender for Movielens, using both implicit (movie watches) and explicit signals (ratings).

## Imports


Let's first get our imports out of the way.


In [2]:
!pip install -q tensorflow-recommenders
!pip install -q --upgrade tensorflow-datasets

In [3]:
import os
import pprint
import tempfile

from typing import Dict, Text

import numpy as np
import tensorflow as tf
import tensorflow_datasets as tfds

2022-12-14 12:23:34.727681: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory
2022-12-14 12:23:34.727787: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory
2022-12-14 12:23:34.727798: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.


In [4]:
import tensorflow_recommenders as tfrs

## Preparing the dataset

We're going to use the Movielens 100K dataset.

In [5]:
ratings = tfds.load('movielens/100k-ratings', split="train")
movies = tfds.load('movielens/100k-movies', split="train")

# Select the basic features.
ratings = ratings.map(lambda x: {
    "movie_title": x["movie_title"],
    "user_id": x["user_id"],
    "user_rating": x["user_rating"],
})
movies = movies.map(lambda x: x["movie_title"])

Instructions for updating:
Lambda fuctions will be no more assumed to be used in the statement where they are used, or at least in the same block. https://github.com/tensorflow/tensorflow/issues/56089


Instructions for updating:
Lambda fuctions will be no more assumed to be used in the statement where they are used, or at least in the same block. https://github.com/tensorflow/tensorflow/issues/56089


And repeat our preparations for building vocabularies and splitting the data into a train and a test set:

In [6]:
# Randomly shuffle data and split between train and test.
tf.random.set_seed(42)
shuffled = ratings.shuffle(100_000, seed=42, reshuffle_each_iteration=False)

train = shuffled.take(80_000)
test = shuffled.skip(80_000).take(20_000)

movie_titles = movies.batch(1_000)
user_ids = ratings.batch(1_000_000).map(lambda x: x["user_id"])

unique_movie_titles = np.unique(np.concatenate(list(movie_titles)))
unique_user_ids = np.unique(np.concatenate(list(user_ids)))

## A multi-task model

There are two critical parts to multi-task recommenders:

1. They optimize for two or more objectives, and so have two or more losses.
2. They share variables between the tasks, allowing for transfer learning.

In this tutorial, we will define our models as before, but instead of having  a single task, we will have two tasks: one that predicts ratings, and one that predicts movie watches.

The user and movie models are as before:

```python
user_model = tf.keras.Sequential([
  tf.keras.layers.StringLookup(
      vocabulary=unique_user_ids, mask_token=None),
  # We add 1 to account for the unknown token.
  tf.keras.layers.Embedding(len(unique_user_ids) + 1, embedding_dimension)
])

movie_model = tf.keras.Sequential([
  tf.keras.layers.StringLookup(
      vocabulary=unique_movie_titles, mask_token=None),
  tf.keras.layers.Embedding(len(unique_movie_titles) + 1, embedding_dimension)
])
```

However, now we will have two tasks. The first is the rating task:

```python
tfrs.tasks.Ranking(
    loss=tf.keras.losses.MeanSquaredError(),
    metrics=[tf.keras.metrics.RootMeanSquaredError()],
)
```

Its goal is to predict the ratings as accurately as possible.

The second is the retrieval task:

```python
tfrs.tasks.Retrieval(
    metrics=tfrs.metrics.FactorizedTopK(
        candidates=movies.batch(128)
    )
)
```

As before, this task's goal is to predict which movies the user will or will not watch.

### Putting it together

We put it all together in a model class.

The new component here is that - since we have two tasks and two losses - we need to decide on how important each loss is. We can do this by giving each of the losses a weight, and treating these weights as hyperparameters. If we assign a large loss weight to the rating task, our model is going to focus on predicting ratings (but still use some information from the retrieval task); if we assign a large loss weight to the retrieval task, it will focus on retrieval instead.

In [7]:
class MovielensModel(tfrs.models.Model):

  def __init__(self, rating_weight: float, retrieval_weight: float) -> None:
    # We take the loss weights in the constructor: this allows us to instantiate
    # several model objects with different loss weights.

    super().__init__()

    embedding_dimension = 32

    # User and movie models.
    self.movie_model: tf.keras.layers.Layer = tf.keras.Sequential([
      tf.keras.layers.StringLookup(
        vocabulary=unique_movie_titles, mask_token=None),
      tf.keras.layers.Embedding(len(unique_movie_titles) + 1, embedding_dimension)
    ])
    self.user_model: tf.keras.layers.Layer = tf.keras.Sequential([
      tf.keras.layers.StringLookup(
        vocabulary=unique_user_ids, mask_token=None),
      tf.keras.layers.Embedding(len(unique_user_ids) + 1, embedding_dimension)
    ])

    # A small model to take in user and movie embeddings and predict ratings.
    # We can make this as complicated as we want as long as we output a scalar
    # as our prediction.
    self.rating_model = tf.keras.Sequential([
        tf.keras.layers.Dense(256, activation="relu"),
        tf.keras.layers.Dense(128, activation="relu"),
        tf.keras.layers.Dense(1),
    ])

    # The tasks.
    self.rating_task: tf.keras.layers.Layer = tfrs.tasks.Ranking(
        loss=tf.keras.losses.MeanSquaredError(),
        metrics=[tf.keras.metrics.RootMeanSquaredError()],
    )
    self.retrieval_task: tf.keras.layers.Layer = tfrs.tasks.Retrieval(
        metrics=tfrs.metrics.FactorizedTopK(
            candidates=movies.batch(128).map(self.movie_model)
        )
    )

    # The loss weights.
    self.rating_weight = rating_weight
    self.retrieval_weight = retrieval_weight

  def call(self, features: Dict[Text, tf.Tensor]) -> tf.Tensor:
    # We pick out the user features and pass them into the user model.
    user_embeddings = self.user_model(features["user_id"])
    # And pick out the movie features and pass them into the movie model.
    movie_embeddings = self.movie_model(features["movie_title"])
    
    return (
        user_embeddings,
        movie_embeddings,
        # We apply the multi-layered rating model to a concatentation of
        # user and movie embeddings.
        self.rating_model(
            tf.concat([user_embeddings, movie_embeddings], axis=1)
        ),
    )

  def compute_loss(self, features: Dict[Text, tf.Tensor], training=False) -> tf.Tensor:

    ratings = features.pop("user_rating")

    user_embeddings, movie_embeddings, rating_predictions = self(features)

    # We compute the loss for each task.
    rating_loss = self.rating_task(
        labels=ratings,
        predictions=rating_predictions,
    )
    retrieval_loss = self.retrieval_task(user_embeddings, movie_embeddings)

    # And combine them using the loss weights.
    return (self.rating_weight * rating_loss
            + self.retrieval_weight * retrieval_loss)

### Rating-specialized model

Depending on the weights we assign, the model will encode a different balance of the tasks. Let's start with a model that only considers ratings.

In [8]:
model = MovielensModel(rating_weight=1.0, retrieval_weight=0.0)
model.compile(optimizer=tf.keras.optimizers.Adagrad(0.1))

In [9]:
cached_train = train.shuffle(100_000).batch(8192).cache()
cached_test = test.batch(4096).cache()

In [10]:
model.fit(cached_train, epochs=3)
metrics = model.evaluate(cached_test, return_dict=True)

print(f"Retrieval top-100 accuracy: {metrics['factorized_top_k/top_100_categorical_accuracy']:.3f}.")
print(f"Ranking RMSE: {metrics['root_mean_squared_error']:.3f}.")

Epoch 1/3


 1/10 [==>...........................] - ETA: 41s - root_mean_squared_error: 3.6920 - factorized_top_k/top_1_categorical_accuracy: 3.6621e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0021 - factorized_top_k/top_10_categorical_accuracy: 0.0054 - factorized_top_k/top_50_categorical_accuracy: 0.0297 - factorized_top_k/top_100_categorical_accuracy: 0.0596 - loss: 13.6308 - regularization_loss: 0.0000e+00 - total_loss: 13.6308

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 3.2088 - factorized_top_k/top_1_categorical_accuracy: 3.0518e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0023 - factorized_top_k/top_10_categorical_accuracy: 0.0062 - factorized_top_k/top_50_categorical_accuracy: 0.0303 - factorized_top_k/top_100_categorical_accuracy: 0.0585 - loss: 10.2962 - regularization_loss: 0.0000e+00 - total_loss: 10.2962 

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 2.9381 - factorized_top_k/top_1_categorical_accuracy: 3.2552e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0063 - factorized_top_k/top_50_categorical_accuracy: 0.0308 - factorized_top_k/top_100_categorical_accuracy: 0.0600 - loss: 8.6325 - regularization_loss: 0.0000e+00 - total_loss: 8.6325  

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 2.9934 - factorized_top_k/top_1_categorical_accuracy: 3.3569e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0067 - factorized_top_k/top_50_categorical_accuracy: 0.0312 - factorized_top_k/top_100_categorical_accuracy: 0.0605 - loss: 8.9603 - regularization_loss: 0.0000e+00 - total_loss: 8.9603

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 2.8774 - factorized_top_k/top_1_categorical_accuracy: 3.1738e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0064 - factorized_top_k/top_50_categorical_accuracy: 0.0314 - factorized_top_k/top_100_categorical_accuracy: 0.0608 - loss: 8.2795 - regularization_loss: 0.0000e+00 - total_loss: 8.2795

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 2.7075 - factorized_top_k/top_1_categorical_accuracy: 3.2552e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0028 - factorized_top_k/top_10_categorical_accuracy: 0.0064 - factorized_top_k/top_50_categorical_accuracy: 0.0313 - factorized_top_k/top_100_categorical_accuracy: 0.0606 - loss: 7.3303 - regularization_loss: 0.0000e+00 - total_loss: 7.3303

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 2.5426 - factorized_top_k/top_1_categorical_accuracy: 3.6621e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0062 - factorized_top_k/top_50_categorical_accuracy: 0.0314 - factorized_top_k/top_100_categorical_accuracy: 0.0607 - loss: 6.4647 - regularization_loss: 0.0000e+00 - total_loss: 6.4647

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 2.4116 - factorized_top_k/top_1_categorical_accuracy: 3.3569e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0060 - factorized_top_k/top_50_categorical_accuracy: 0.0308 - factorized_top_k/top_100_categorical_accuracy: 0.0601 - loss: 5.8160 - regularization_loss: 0.0000e+00 - total_loss: 5.8160

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 2.3049 - factorized_top_k/top_1_categorical_accuracy: 3.5265e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0060 - factorized_top_k/top_50_categorical_accuracy: 0.0307 - factorized_top_k/top_100_categorical_accuracy: 0.0598 - loss: 5.3123 - regularization_loss: 0.0000e+00 - total_loss: 5.3123

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 2.2354 - factorized_top_k/top_1_categorical_accuracy: 3.3750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0060 - factorized_top_k/top_50_categorical_accuracy: 0.0305 - factorized_top_k/top_100_categorical_accuracy: 0.0599 - loss: 4.9100 - regularization_loss: 0.0000e+00 - total_loss: 4.9100

10/10 [==============================] - 7s 319ms/step - root_mean_squared_error: 2.2354 - factorized_top_k/top_1_categorical_accuracy: 3.3750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0060 - factorized_top_k/top_50_categorical_accuracy: 0.0305 - factorized_top_k/top_100_categorical_accuracy: 0.0599 - loss: 4.5809 - regularization_loss: 0.0000e+00 - total_loss: 4.5809


Epoch 2/3


 1/10 [==>...........................] - ETA: 2s - root_mean_squared_error: 1.1171 - factorized_top_k/top_1_categorical_accuracy: 3.6621e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0021 - factorized_top_k/top_10_categorical_accuracy: 0.0051 - factorized_top_k/top_50_categorical_accuracy: 0.0297 - factorized_top_k/top_100_categorical_accuracy: 0.0596 - loss: 1.2478 - regularization_loss: 0.0000e+00 - total_loss: 1.2478

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 1.1216 - factorized_top_k/top_1_categorical_accuracy: 3.0518e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0022 - factorized_top_k/top_10_categorical_accuracy: 0.0059 - factorized_top_k/top_50_categorical_accuracy: 0.0301 - factorized_top_k/top_100_categorical_accuracy: 0.0586 - loss: 1.2581 - regularization_loss: 0.0000e+00 - total_loss: 1.2581

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 1.1259 - factorized_top_k/top_1_categorical_accuracy: 2.4414e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0024 - factorized_top_k/top_10_categorical_accuracy: 0.0059 - factorized_top_k/top_50_categorical_accuracy: 0.0306 - factorized_top_k/top_100_categorical_accuracy: 0.0603 - loss: 1.2675 - regularization_loss: 0.0000e+00 - total_loss: 1.2675

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 1.1240 - factorized_top_k/top_1_categorical_accuracy: 2.7466e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0062 - factorized_top_k/top_50_categorical_accuracy: 0.0311 - factorized_top_k/top_100_categorical_accuracy: 0.0609 - loss: 1.2635 - regularization_loss: 0.0000e+00 - total_loss: 1.2635

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 1.1226 - factorized_top_k/top_1_categorical_accuracy: 2.6855e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0060 - factorized_top_k/top_50_categorical_accuracy: 0.0312 - factorized_top_k/top_100_categorical_accuracy: 0.0613 - loss: 1.2603 - regularization_loss: 0.0000e+00 - total_loss: 1.2603

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 1.1211 - factorized_top_k/top_1_categorical_accuracy: 2.8483e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0060 - factorized_top_k/top_50_categorical_accuracy: 0.0311 - factorized_top_k/top_100_categorical_accuracy: 0.0612 - loss: 1.2570 - regularization_loss: 0.0000e+00 - total_loss: 1.2570

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 1.1201 - factorized_top_k/top_1_categorical_accuracy: 2.7902e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0058 - factorized_top_k/top_50_categorical_accuracy: 0.0313 - factorized_top_k/top_100_categorical_accuracy: 0.0613 - loss: 1.2545 - regularization_loss: 0.0000e+00 - total_loss: 1.2545

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.1203 - factorized_top_k/top_1_categorical_accuracy: 2.5940e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0024 - factorized_top_k/top_10_categorical_accuracy: 0.0056 - factorized_top_k/top_50_categorical_accuracy: 0.0307 - factorized_top_k/top_100_categorical_accuracy: 0.0606 - loss: 1.2552 - regularization_loss: 0.0000e+00 - total_loss: 1.2552

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 1.1212 - factorized_top_k/top_1_categorical_accuracy: 2.5770e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0056 - factorized_top_k/top_50_categorical_accuracy: 0.0306 - factorized_top_k/top_100_categorical_accuracy: 0.0602 - loss: 1.2571 - regularization_loss: 0.0000e+00 - total_loss: 1.2571

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 1.1220 - factorized_top_k/top_1_categorical_accuracy: 2.6250e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0056 - factorized_top_k/top_50_categorical_accuracy: 0.0304 - factorized_top_k/top_100_categorical_accuracy: 0.0601 - loss: 1.2595 - regularization_loss: 0.0000e+00 - total_loss: 1.2595

10/10 [==============================] - 3s 319ms/step - root_mean_squared_error: 1.1220 - factorized_top_k/top_1_categorical_accuracy: 2.6250e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0056 - factorized_top_k/top_50_categorical_accuracy: 0.0304 - factorized_top_k/top_100_categorical_accuracy: 0.0601 - loss: 1.2614 - regularization_loss: 0.0000e+00 - total_loss: 1.2614


Epoch 3/3


 1/10 [==>...........................] - ETA: 2s - root_mean_squared_error: 1.1127 - factorized_top_k/top_1_categorical_accuracy: 2.4414e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0021 - factorized_top_k/top_10_categorical_accuracy: 0.0054 - factorized_top_k/top_50_categorical_accuracy: 0.0303 - factorized_top_k/top_100_categorical_accuracy: 0.0594 - loss: 1.2382 - regularization_loss: 0.0000e+00 - total_loss: 1.2382

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 1.1171 - factorized_top_k/top_1_categorical_accuracy: 2.4414e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0020 - factorized_top_k/top_10_categorical_accuracy: 0.0059 - factorized_top_k/top_50_categorical_accuracy: 0.0302 - factorized_top_k/top_100_categorical_accuracy: 0.0587 - loss: 1.2480 - regularization_loss: 0.0000e+00 - total_loss: 1.2480

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 1.1213 - factorized_top_k/top_1_categorical_accuracy: 2.0345e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0023 - factorized_top_k/top_10_categorical_accuracy: 0.0059 - factorized_top_k/top_50_categorical_accuracy: 0.0306 - factorized_top_k/top_100_categorical_accuracy: 0.0603 - loss: 1.2573 - regularization_loss: 0.0000e+00 - total_loss: 1.2573

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 1.1194 - factorized_top_k/top_1_categorical_accuracy: 2.4414e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0024 - factorized_top_k/top_10_categorical_accuracy: 0.0063 - factorized_top_k/top_50_categorical_accuracy: 0.0310 - factorized_top_k/top_100_categorical_accuracy: 0.0610 - loss: 1.2531 - regularization_loss: 0.0000e+00 - total_loss: 1.2531

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 1.1179 - factorized_top_k/top_1_categorical_accuracy: 2.4414e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0061 - factorized_top_k/top_50_categorical_accuracy: 0.0311 - factorized_top_k/top_100_categorical_accuracy: 0.0618 - loss: 1.2497 - regularization_loss: 0.0000e+00 - total_loss: 1.2497

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 1.1164 - factorized_top_k/top_1_categorical_accuracy: 2.8483e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0061 - factorized_top_k/top_50_categorical_accuracy: 0.0312 - factorized_top_k/top_100_categorical_accuracy: 0.0615 - loss: 1.2463 - regularization_loss: 0.0000e+00 - total_loss: 1.2463

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 1.1152 - factorized_top_k/top_1_categorical_accuracy: 2.7902e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0025 - factorized_top_k/top_10_categorical_accuracy: 0.0059 - factorized_top_k/top_50_categorical_accuracy: 0.0313 - factorized_top_k/top_100_categorical_accuracy: 0.0617 - loss: 1.2437 - regularization_loss: 0.0000e+00 - total_loss: 1.2437

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.1154 - factorized_top_k/top_1_categorical_accuracy: 2.5940e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0024 - factorized_top_k/top_10_categorical_accuracy: 0.0057 - factorized_top_k/top_50_categorical_accuracy: 0.0306 - factorized_top_k/top_100_categorical_accuracy: 0.0610 - loss: 1.2442 - regularization_loss: 0.0000e+00 - total_loss: 1.2442

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 1.1162 - factorized_top_k/top_1_categorical_accuracy: 2.5770e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0024 - factorized_top_k/top_10_categorical_accuracy: 0.0057 - factorized_top_k/top_50_categorical_accuracy: 0.0305 - factorized_top_k/top_100_categorical_accuracy: 0.0606 - loss: 1.2460 - regularization_loss: 0.0000e+00 - total_loss: 1.2460

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 1.1170 - factorized_top_k/top_1_categorical_accuracy: 2.6250e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0024 - factorized_top_k/top_10_categorical_accuracy: 0.0057 - factorized_top_k/top_50_categorical_accuracy: 0.0304 - factorized_top_k/top_100_categorical_accuracy: 0.0605 - loss: 1.2482 - regularization_loss: 0.0000e+00 - total_loss: 1.2482

10/10 [==============================] - 3s 315ms/step - root_mean_squared_error: 1.1170 - factorized_top_k/top_1_categorical_accuracy: 2.6250e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0024 - factorized_top_k/top_10_categorical_accuracy: 0.0057 - factorized_top_k/top_50_categorical_accuracy: 0.0304 - factorized_top_k/top_100_categorical_accuracy: 0.0605 - loss: 1.2500 - regularization_loss: 0.0000e+00 - total_loss: 1.2500


1/5 [=====>........................] - ETA: 7s - root_mean_squared_error: 1.1162 - factorized_top_k/top_1_categorical_accuracy: 4.8828e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0032 - factorized_top_k/top_10_categorical_accuracy: 0.0054 - factorized_top_k/top_50_categorical_accuracy: 0.0254 - factorized_top_k/top_100_categorical_accuracy: 0.0605 - loss: 1.2459 - regularization_loss: 0.0000e+00 - total_loss: 1.2459

2/5 [===========>..................] - ETA: 0s - root_mean_squared_error: 1.1202 - factorized_top_k/top_1_categorical_accuracy: 3.6621e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0029 - factorized_top_k/top_10_categorical_accuracy: 0.0060 - factorized_top_k/top_50_categorical_accuracy: 0.0308 - factorized_top_k/top_100_categorical_accuracy: 0.0616 - loss: 1.2548 - regularization_loss: 0.0000e+00 - total_loss: 1.2548

3/5 [=================>............] - ETA: 0s - root_mean_squared_error: 1.1194 - factorized_top_k/top_1_categorical_accuracy: 2.4414e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0027 - factorized_top_k/top_10_categorical_accuracy: 0.0051 - factorized_top_k/top_50_categorical_accuracy: 0.0288 - factorized_top_k/top_100_categorical_accuracy: 0.0582 - loss: 1.2531 - regularization_loss: 0.0000e+00 - total_loss: 1.2531

4/5 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.1153 - factorized_top_k/top_1_categorical_accuracy: 4.2725e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0034 - factorized_top_k/top_10_categorical_accuracy: 0.0062 - factorized_top_k/top_50_categorical_accuracy: 0.0303 - factorized_top_k/top_100_categorical_accuracy: 0.0599 - loss: 1.2439 - regularization_loss: 0.0000e+00 - total_loss: 1.2439

5/5 [==============================] - ETA: 0s - root_mean_squared_error: 1.1125 - factorized_top_k/top_1_categorical_accuracy: 5.0000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0034 - factorized_top_k/top_10_categorical_accuracy: 0.0065 - factorized_top_k/top_50_categorical_accuracy: 0.0309 - factorized_top_k/top_100_categorical_accuracy: 0.0599 - loss: 1.2371 - regularization_loss: 0.0000e+00 - total_loss: 1.2371

5/5 [==============================] - 3s 185ms/step - root_mean_squared_error: 1.1125 - factorized_top_k/top_1_categorical_accuracy: 5.0000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0034 - factorized_top_k/top_10_categorical_accuracy: 0.0065 - factorized_top_k/top_50_categorical_accuracy: 0.0309 - factorized_top_k/top_100_categorical_accuracy: 0.0599 - loss: 1.2326 - regularization_loss: 0.0000e+00 - total_loss: 1.2326


Retrieval top-100 accuracy: 0.060.
Ranking RMSE: 1.113.


The model does OK on predicting ratings (with an RMSE of around 1.11), but performs poorly at predicting which movies will be watched or not: its accuracy at 100 is almost 4 times worse than a model trained solely to predict watches.

### Retrieval-specialized model

Let's now try a model that focuses on retrieval only.

In [11]:
model = MovielensModel(rating_weight=0.0, retrieval_weight=1.0)
model.compile(optimizer=tf.keras.optimizers.Adagrad(0.1))

In [12]:
model.fit(cached_train, epochs=3)
metrics = model.evaluate(cached_test, return_dict=True)

print(f"Retrieval top-100 accuracy: {metrics['factorized_top_k/top_100_categorical_accuracy']:.3f}.")
print(f"Ranking RMSE: {metrics['root_mean_squared_error']:.3f}.")

Epoch 1/3


 1/10 [==>...........................] - ETA: 14s - root_mean_squared_error: 3.6977 - factorized_top_k/top_1_categorical_accuracy: 4.8828e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0020 - factorized_top_k/top_10_categorical_accuracy: 0.0048 - factorized_top_k/top_50_categorical_accuracy: 0.0299 - factorized_top_k/top_100_categorical_accuracy: 0.0602 - loss: 73818.1250 - regularization_loss: 0.0000e+00 - total_loss: 73818.1250

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 3.7078 - factorized_top_k/top_1_categorical_accuracy: 7.3242e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0044 - factorized_top_k/top_10_categorical_accuracy: 0.0096 - factorized_top_k/top_50_categorical_accuracy: 0.0409 - factorized_top_k/top_100_categorical_accuracy: 0.0742 - loss: 73819.0781 - regularization_loss: 0.0000e+00 - total_loss: 73819.0781 

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 3.7113 - factorized_top_k/top_1_categorical_accuracy: 6.9173e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0048 - factorized_top_k/top_10_categorical_accuracy: 0.0105 - factorized_top_k/top_50_categorical_accuracy: 0.0470 - factorized_top_k/top_100_categorical_accuracy: 0.0834 - loss: 73820.5833 - regularization_loss: 0.0000e+00 - total_loss: 73820.5833

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 3.7060 - factorized_top_k/top_1_categorical_accuracy: 6.1035e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0051 - factorized_top_k/top_10_categorical_accuracy: 0.0117 - factorized_top_k/top_50_categorical_accuracy: 0.0525 - factorized_top_k/top_100_categorical_accuracy: 0.0935 - loss: 73818.0391 - regularization_loss: 0.0000e+00 - total_loss: 73818.0391

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 3.7024 - factorized_top_k/top_1_categorical_accuracy: 6.5918e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0059 - factorized_top_k/top_10_categorical_accuracy: 0.0132 - factorized_top_k/top_50_categorical_accuracy: 0.0614 - factorized_top_k/top_100_categorical_accuracy: 0.1099 - loss: 73803.2734 - regularization_loss: 0.0000e+00 - total_loss: 73803.2734

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 3.7012 - factorized_top_k/top_1_categorical_accuracy: 7.5277e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0068 - factorized_top_k/top_10_categorical_accuracy: 0.0153 - factorized_top_k/top_50_categorical_accuracy: 0.0723 - factorized_top_k/top_100_categorical_accuracy: 0.1286 - loss: 73765.4115 - regularization_loss: 0.0000e+00 - total_loss: 73765.4115

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 3.7009 - factorized_top_k/top_1_categorical_accuracy: 6.4523e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0060 - factorized_top_k/top_10_categorical_accuracy: 0.0136 - factorized_top_k/top_50_categorical_accuracy: 0.0678 - factorized_top_k/top_100_categorical_accuracy: 0.1276 - loss: 73693.1875 - regularization_loss: 0.0000e+00 - total_loss: 73693.1875

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 3.6993 - factorized_top_k/top_1_categorical_accuracy: 5.7983e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0053 - factorized_top_k/top_10_categorical_accuracy: 0.0123 - factorized_top_k/top_50_categorical_accuracy: 0.0668 - factorized_top_k/top_100_categorical_accuracy: 0.1320 - loss: 73586.3799 - regularization_loss: 0.0000e+00 - total_loss: 73586.3799

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 3.6966 - factorized_top_k/top_1_categorical_accuracy: 5.1541e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0049 - factorized_top_k/top_10_categorical_accuracy: 0.0116 - factorized_top_k/top_50_categorical_accuracy: 0.0681 - factorized_top_k/top_100_categorical_accuracy: 0.1373 - loss: 73459.6832 - regularization_loss: 0.0000e+00 - total_loss: 73459.6832

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 3.6972 - factorized_top_k/top_1_categorical_accuracy: 5.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0056 - factorized_top_k/top_10_categorical_accuracy: 0.0131 - factorized_top_k/top_50_categorical_accuracy: 0.0751 - factorized_top_k/top_100_categorical_accuracy: 0.1483 - loss: 71462.8961 - regularization_loss: 0.0000e+00 - total_loss: 71462.8961

10/10 [==============================] - 4s 309ms/step - root_mean_squared_error: 3.6972 - factorized_top_k/top_1_categorical_accuracy: 5.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0056 - factorized_top_k/top_10_categorical_accuracy: 0.0131 - factorized_top_k/top_50_categorical_accuracy: 0.0751 - factorized_top_k/top_100_categorical_accuracy: 0.1483 - loss: 69829.1612 - regularization_loss: 0.0000e+00 - total_loss: 69829.1612


Epoch 2/3


 1/10 [==>...........................] - ETA: 2s - root_mean_squared_error: 3.6878 - factorized_top_k/top_1_categorical_accuracy: 2.4414e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0055 - factorized_top_k/top_10_categorical_accuracy: 0.0140 - factorized_top_k/top_50_categorical_accuracy: 0.1040 - factorized_top_k/top_100_categorical_accuracy: 0.2114 - loss: 71794.9609 - regularization_loss: 0.0000e+00 - total_loss: 71794.9609

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 3.6989 - factorized_top_k/top_1_categorical_accuracy: 6.1035e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0084 - factorized_top_k/top_10_categorical_accuracy: 0.0181 - factorized_top_k/top_50_categorical_accuracy: 0.1132 - factorized_top_k/top_100_categorical_accuracy: 0.2273 - loss: 71635.9883 - regularization_loss: 0.0000e+00 - total_loss: 71635.9883

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 3.7035 - factorized_top_k/top_1_categorical_accuracy: 6.5104e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0087 - factorized_top_k/top_10_categorical_accuracy: 0.0193 - factorized_top_k/top_50_categorical_accuracy: 0.1201 - factorized_top_k/top_100_categorical_accuracy: 0.2361 - loss: 71512.4141 - regularization_loss: 0.0000e+00 - total_loss: 71512.4141

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 3.6983 - factorized_top_k/top_1_categorical_accuracy: 9.1553e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0101 - factorized_top_k/top_10_categorical_accuracy: 0.0219 - factorized_top_k/top_50_categorical_accuracy: 0.1252 - factorized_top_k/top_100_categorical_accuracy: 0.2429 - loss: 71387.2773 - regularization_loss: 0.0000e+00 - total_loss: 71387.2773

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 3.6948 - factorized_top_k/top_1_categorical_accuracy: 8.0566e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0104 - factorized_top_k/top_10_categorical_accuracy: 0.0229 - factorized_top_k/top_50_categorical_accuracy: 0.1296 - factorized_top_k/top_100_categorical_accuracy: 0.2485 - loss: 71253.5281 - regularization_loss: 0.0000e+00 - total_loss: 71253.5281

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 3.6935 - factorized_top_k/top_1_categorical_accuracy: 0.0010 - factorized_top_k/top_5_categorical_accuracy: 0.0108 - factorized_top_k/top_10_categorical_accuracy: 0.0240 - factorized_top_k/top_50_categorical_accuracy: 0.1334 - factorized_top_k/top_100_categorical_accuracy: 0.2530 - loss: 71144.3841 - regularization_loss: 0.0000e+00 - total_loss: 71144.3841    

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 3.6932 - factorized_top_k/top_1_categorical_accuracy: 0.0010 - factorized_top_k/top_5_categorical_accuracy: 0.0112 - factorized_top_k/top_10_categorical_accuracy: 0.0250 - factorized_top_k/top_50_categorical_accuracy: 0.1360 - factorized_top_k/top_100_categorical_accuracy: 0.2558 - loss: 71051.0257 - regularization_loss: 0.0000e+00 - total_loss: 71051.0257

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 3.6920 - factorized_top_k/top_1_categorical_accuracy: 9.6130e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0113 - factorized_top_k/top_10_categorical_accuracy: 0.0259 - factorized_top_k/top_50_categorical_accuracy: 0.1388 - factorized_top_k/top_100_categorical_accuracy: 0.2584 - loss: 70973.2822 - regularization_loss: 0.0000e+00 - total_loss: 70973.2822

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 3.6895 - factorized_top_k/top_1_categorical_accuracy: 0.0010 - factorized_top_k/top_5_categorical_accuracy: 0.0116 - factorized_top_k/top_10_categorical_accuracy: 0.0268 - factorized_top_k/top_50_categorical_accuracy: 0.1410 - factorized_top_k/top_100_categorical_accuracy: 0.2609 - loss: 70898.9158 - regularization_loss: 0.0000e+00 - total_loss: 70898.9158    

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 3.6905 - factorized_top_k/top_1_categorical_accuracy: 0.0010 - factorized_top_k/top_5_categorical_accuracy: 0.0118 - factorized_top_k/top_10_categorical_accuracy: 0.0272 - factorized_top_k/top_50_categorical_accuracy: 0.1425 - factorized_top_k/top_100_categorical_accuracy: 0.2634 - loss: 69010.8484 - regularization_loss: 0.0000e+00 - total_loss: 69010.8484

10/10 [==============================] - 3s 301ms/step - root_mean_squared_error: 3.6905 - factorized_top_k/top_1_categorical_accuracy: 0.0010 - factorized_top_k/top_5_categorical_accuracy: 0.0118 - factorized_top_k/top_10_categorical_accuracy: 0.0272 - factorized_top_k/top_50_categorical_accuracy: 0.1425 - factorized_top_k/top_100_categorical_accuracy: 0.2634 - loss: 67466.0661 - regularization_loss: 0.0000e+00 - total_loss: 67466.0661


Epoch 3/3


 1/10 [==>...........................] - ETA: 2s - root_mean_squared_error: 3.6859 - factorized_top_k/top_1_categorical_accuracy: 6.1035e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0137 - factorized_top_k/top_10_categorical_accuracy: 0.0288 - factorized_top_k/top_50_categorical_accuracy: 0.1566 - factorized_top_k/top_100_categorical_accuracy: 0.2805 - loss: 70024.9688 - regularization_loss: 0.0000e+00 - total_loss: 70024.9688

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 3.6969 - factorized_top_k/top_1_categorical_accuracy: 0.0016 - factorized_top_k/top_5_categorical_accuracy: 0.0161 - factorized_top_k/top_10_categorical_accuracy: 0.0332 - factorized_top_k/top_50_categorical_accuracy: 0.1632 - factorized_top_k/top_100_categorical_accuracy: 0.2882 - loss: 69953.2734 - regularization_loss: 0.0000e+00 - total_loss: 69953.2734    

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 3.7013 - factorized_top_k/top_1_categorical_accuracy: 0.0016 - factorized_top_k/top_5_categorical_accuracy: 0.0166 - factorized_top_k/top_10_categorical_accuracy: 0.0343 - factorized_top_k/top_50_categorical_accuracy: 0.1680 - factorized_top_k/top_100_categorical_accuracy: 0.2941 - loss: 69889.3750 - regularization_loss: 0.0000e+00 - total_loss: 69889.3750

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 3.6960 - factorized_top_k/top_1_categorical_accuracy: 0.0016 - factorized_top_k/top_5_categorical_accuracy: 0.0175 - factorized_top_k/top_10_categorical_accuracy: 0.0367 - factorized_top_k/top_50_categorical_accuracy: 0.1719 - factorized_top_k/top_100_categorical_accuracy: 0.2973 - loss: 69834.5547 - regularization_loss: 0.0000e+00 - total_loss: 69834.5547

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 3.6924 - factorized_top_k/top_1_categorical_accuracy: 0.0016 - factorized_top_k/top_5_categorical_accuracy: 0.0176 - factorized_top_k/top_10_categorical_accuracy: 0.0374 - factorized_top_k/top_50_categorical_accuracy: 0.1742 - factorized_top_k/top_100_categorical_accuracy: 0.2994 - loss: 69766.0406 - regularization_loss: 0.0000e+00 - total_loss: 69766.0406

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 3.6910 - factorized_top_k/top_1_categorical_accuracy: 0.0018 - factorized_top_k/top_5_categorical_accuracy: 0.0180 - factorized_top_k/top_10_categorical_accuracy: 0.0379 - factorized_top_k/top_50_categorical_accuracy: 0.1755 - factorized_top_k/top_100_categorical_accuracy: 0.3017 - loss: 69712.2747 - regularization_loss: 0.0000e+00 - total_loss: 69712.2747

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 3.6906 - factorized_top_k/top_1_categorical_accuracy: 0.0017 - factorized_top_k/top_5_categorical_accuracy: 0.0182 - factorized_top_k/top_10_categorical_accuracy: 0.0381 - factorized_top_k/top_50_categorical_accuracy: 0.1755 - factorized_top_k/top_100_categorical_accuracy: 0.3027 - loss: 69670.3114 - regularization_loss: 0.0000e+00 - total_loss: 69670.3114

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 3.6893 - factorized_top_k/top_1_categorical_accuracy: 0.0016 - factorized_top_k/top_5_categorical_accuracy: 0.0180 - factorized_top_k/top_10_categorical_accuracy: 0.0382 - factorized_top_k/top_50_categorical_accuracy: 0.1767 - factorized_top_k/top_100_categorical_accuracy: 0.3036 - loss: 69641.3213 - regularization_loss: 0.0000e+00 - total_loss: 69641.3213

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 3.6868 - factorized_top_k/top_1_categorical_accuracy: 0.0017 - factorized_top_k/top_5_categorical_accuracy: 0.0182 - factorized_top_k/top_10_categorical_accuracy: 0.0387 - factorized_top_k/top_50_categorical_accuracy: 0.1773 - factorized_top_k/top_100_categorical_accuracy: 0.3041 - loss: 69612.8620 - regularization_loss: 0.0000e+00 - total_loss: 69612.8620

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 3.6877 - factorized_top_k/top_1_categorical_accuracy: 0.0016 - factorized_top_k/top_5_categorical_accuracy: 0.0183 - factorized_top_k/top_10_categorical_accuracy: 0.0391 - factorized_top_k/top_50_categorical_accuracy: 0.1782 - factorized_top_k/top_100_categorical_accuracy: 0.3048 - loss: 67787.7699 - regularization_loss: 0.0000e+00 - total_loss: 67787.7699

10/10 [==============================] - 3s 300ms/step - root_mean_squared_error: 3.6877 - factorized_top_k/top_1_categorical_accuracy: 0.0016 - factorized_top_k/top_5_categorical_accuracy: 0.0183 - factorized_top_k/top_10_categorical_accuracy: 0.0391 - factorized_top_k/top_50_categorical_accuracy: 0.1782 - factorized_top_k/top_100_categorical_accuracy: 0.3048 - loss: 66294.5128 - regularization_loss: 0.0000e+00 - total_loss: 66294.5128


1/5 [=====>........................] - ETA: 2s - root_mean_squared_error: 3.6898 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0081 - factorized_top_k/top_10_categorical_accuracy: 0.0195 - factorized_top_k/top_50_categorical_accuracy: 0.1238 - factorized_top_k/top_100_categorical_accuracy: 0.2363 - loss: 32480.8496 - regularization_loss: 0.0000e+00 - total_loss: 32480.8496

2/5 [===========>..................] - ETA: 0s - root_mean_squared_error: 3.6918 - factorized_top_k/top_1_categorical_accuracy: 9.7656e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0082 - factorized_top_k/top_10_categorical_accuracy: 0.0182 - factorized_top_k/top_50_categorical_accuracy: 0.1184 - factorized_top_k/top_100_categorical_accuracy: 0.2306 - loss: 32527.2363 - regularization_loss: 0.0000e+00 - total_loss: 32527.2363

3/5 [=================>............] - ETA: 0s - root_mean_squared_error: 3.6907 - factorized_top_k/top_1_categorical_accuracy: 9.7656e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0094 - factorized_top_k/top_10_categorical_accuracy: 0.0197 - factorized_top_k/top_50_categorical_accuracy: 0.1186 - factorized_top_k/top_100_categorical_accuracy: 0.2327 - loss: 32526.7005 - regularization_loss: 0.0000e+00 - total_loss: 32526.7005

4/5 [=======================>......] - ETA: 0s - root_mean_squared_error: 3.6890 - factorized_top_k/top_1_categorical_accuracy: 9.1553e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0088 - factorized_top_k/top_10_categorical_accuracy: 0.0195 - factorized_top_k/top_50_categorical_accuracy: 0.1190 - factorized_top_k/top_100_categorical_accuracy: 0.2332 - loss: 32508.8735 - regularization_loss: 0.0000e+00 - total_loss: 32508.8735

5/5 [==============================] - ETA: 0s - root_mean_squared_error: 3.6884 - factorized_top_k/top_1_categorical_accuracy: 9.5000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0093 - factorized_top_k/top_10_categorical_accuracy: 0.0203 - factorized_top_k/top_50_categorical_accuracy: 0.1199 - factorized_top_k/top_100_categorical_accuracy: 0.2330 - loss: 31658.8367 - regularization_loss: 0.0000e+00 - total_loss: 31658.8367

5/5 [==============================] - 1s 188ms/step - root_mean_squared_error: 3.6884 - factorized_top_k/top_1_categorical_accuracy: 9.5000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0093 - factorized_top_k/top_10_categorical_accuracy: 0.0203 - factorized_top_k/top_50_categorical_accuracy: 0.1199 - factorized_top_k/top_100_categorical_accuracy: 0.2330 - loss: 31092.1455 - regularization_loss: 0.0000e+00 - total_loss: 31092.1455


Retrieval top-100 accuracy: 0.233.
Ranking RMSE: 3.688.


We get the opposite result: a model that does well on retrieval, but poorly on predicting ratings.

### Joint model

Let's now train a model that assigns positive weights to both tasks.

In [13]:
model = MovielensModel(rating_weight=1.0, retrieval_weight=1.0)
model.compile(optimizer=tf.keras.optimizers.Adagrad(0.1))

In [14]:
model.fit(cached_train, epochs=3)
metrics = model.evaluate(cached_test, return_dict=True)

print(f"Retrieval top-100 accuracy: {metrics['factorized_top_k/top_100_categorical_accuracy']:.3f}.")
print(f"Ranking RMSE: {metrics['root_mean_squared_error']:.3f}.")

Epoch 1/3


 1/10 [==>...........................] - ETA: 13s - root_mean_squared_error: 3.7058 - factorized_top_k/top_1_categorical_accuracy: 3.6621e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0020 - factorized_top_k/top_10_categorical_accuracy: 0.0054 - factorized_top_k/top_50_categorical_accuracy: 0.0277 - factorized_top_k/top_100_categorical_accuracy: 0.0558 - loss: 73831.4688 - regularization_loss: 0.0000e+00 - total_loss: 73831.4688

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 3.3079 - factorized_top_k/top_1_categorical_accuracy: 6.1035e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0034 - factorized_top_k/top_10_categorical_accuracy: 0.0082 - factorized_top_k/top_50_categorical_accuracy: 0.0374 - factorized_top_k/top_100_categorical_accuracy: 0.0695 - loss: 73829.6445 - regularization_loss: 0.0000e+00 - total_loss: 73829.6445 

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 2.8337 - factorized_top_k/top_1_categorical_accuracy: 4.8828e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0041 - factorized_top_k/top_10_categorical_accuracy: 0.0095 - factorized_top_k/top_50_categorical_accuracy: 0.0455 - factorized_top_k/top_100_categorical_accuracy: 0.0819 - loss: 73827.5000 - regularization_loss: 0.0000e+00 - total_loss: 73827.5000

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 2.7453 - factorized_top_k/top_1_categorical_accuracy: 4.5776e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0045 - factorized_top_k/top_10_categorical_accuracy: 0.0105 - factorized_top_k/top_50_categorical_accuracy: 0.0525 - factorized_top_k/top_100_categorical_accuracy: 0.0930 - loss: 73822.8281 - regularization_loss: 0.0000e+00 - total_loss: 73822.8281

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 2.5060 - factorized_top_k/top_1_categorical_accuracy: 6.1035e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0054 - factorized_top_k/top_10_categorical_accuracy: 0.0127 - factorized_top_k/top_50_categorical_accuracy: 0.0629 - factorized_top_k/top_100_categorical_accuracy: 0.1112 - loss: 73803.5844 - regularization_loss: 0.0000e+00 - total_loss: 73803.5844

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 2.3329 - factorized_top_k/top_1_categorical_accuracy: 6.7139e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0062 - factorized_top_k/top_10_categorical_accuracy: 0.0145 - factorized_top_k/top_50_categorical_accuracy: 0.0739 - factorized_top_k/top_100_categorical_accuracy: 0.1307 - loss: 73758.1992 - regularization_loss: 0.0000e+00 - total_loss: 73758.1992

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 2.2017 - factorized_top_k/top_1_categorical_accuracy: 5.7547e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0053 - factorized_top_k/top_10_categorical_accuracy: 0.0126 - factorized_top_k/top_50_categorical_accuracy: 0.0700 - factorized_top_k/top_100_categorical_accuracy: 0.1307 - loss: 73675.1942 - regularization_loss: 0.0000e+00 - total_loss: 73675.1942

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 2.1047 - factorized_top_k/top_1_categorical_accuracy: 5.0354e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0047 - factorized_top_k/top_10_categorical_accuracy: 0.0115 - factorized_top_k/top_50_categorical_accuracy: 0.0696 - factorized_top_k/top_100_categorical_accuracy: 0.1355 - loss: 73556.0908 - regularization_loss: 0.0000e+00 - total_loss: 73556.0908

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 2.0493 - factorized_top_k/top_1_categorical_accuracy: 4.4759e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0044 - factorized_top_k/top_10_categorical_accuracy: 0.0111 - factorized_top_k/top_50_categorical_accuracy: 0.0713 - factorized_top_k/top_100_categorical_accuracy: 0.1417 - loss: 73421.8837 - regularization_loss: 0.0000e+00 - total_loss: 73421.8837

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 2.0230 - factorized_top_k/top_1_categorical_accuracy: 5.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0051 - factorized_top_k/top_10_categorical_accuracy: 0.0123 - factorized_top_k/top_50_categorical_accuracy: 0.0768 - factorized_top_k/top_100_categorical_accuracy: 0.1509 - loss: 71422.9023 - regularization_loss: 0.0000e+00 - total_loss: 71422.9023

10/10 [==============================] - 4s 309ms/step - root_mean_squared_error: 2.0230 - factorized_top_k/top_1_categorical_accuracy: 5.8750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0051 - factorized_top_k/top_10_categorical_accuracy: 0.0123 - factorized_top_k/top_50_categorical_accuracy: 0.0768 - factorized_top_k/top_100_categorical_accuracy: 0.1509 - loss: 69787.3722 - regularization_loss: 0.0000e+00 - total_loss: 69787.3722


Epoch 2/3


 1/10 [==>...........................] - ETA: 2s - root_mean_squared_error: 2.3682 - factorized_top_k/top_1_categorical_accuracy: 7.3242e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0083 - factorized_top_k/top_10_categorical_accuracy: 0.0161 - factorized_top_k/top_50_categorical_accuracy: 0.1001 - factorized_top_k/top_100_categorical_accuracy: 0.2144 - loss: 71739.6562 - regularization_loss: 0.0000e+00 - total_loss: 71739.6562

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 1.8633 - factorized_top_k/top_1_categorical_accuracy: 0.0010 - factorized_top_k/top_5_categorical_accuracy: 0.0090 - factorized_top_k/top_10_categorical_accuracy: 0.0193 - factorized_top_k/top_50_categorical_accuracy: 0.1142 - factorized_top_k/top_100_categorical_accuracy: 0.2304 - loss: 71578.6289 - regularization_loss: 0.0000e+00 - total_loss: 71578.6289    

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 1.6512 - factorized_top_k/top_1_categorical_accuracy: 9.7656e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0097 - factorized_top_k/top_10_categorical_accuracy: 0.0208 - factorized_top_k/top_50_categorical_accuracy: 0.1220 - factorized_top_k/top_100_categorical_accuracy: 0.2390 - loss: 71464.9714 - regularization_loss: 0.0000e+00 - total_loss: 71464.9714

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 1.5325 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0107 - factorized_top_k/top_10_categorical_accuracy: 0.0228 - factorized_top_k/top_50_categorical_accuracy: 0.1287 - factorized_top_k/top_100_categorical_accuracy: 0.2450 - loss: 71348.3203 - regularization_loss: 0.0000e+00 - total_loss: 71348.3203    

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 1.4593 - factorized_top_k/top_1_categorical_accuracy: 0.0011 - factorized_top_k/top_5_categorical_accuracy: 0.0107 - factorized_top_k/top_10_categorical_accuracy: 0.0235 - factorized_top_k/top_50_categorical_accuracy: 0.1329 - factorized_top_k/top_100_categorical_accuracy: 0.2504 - loss: 71221.7828 - regularization_loss: 0.0000e+00 - total_loss: 71221.7828

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 1.4144 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0114 - factorized_top_k/top_10_categorical_accuracy: 0.0247 - factorized_top_k/top_50_categorical_accuracy: 0.1355 - factorized_top_k/top_100_categorical_accuracy: 0.2543 - loss: 71117.7513 - regularization_loss: 0.0000e+00 - total_loss: 71117.7513

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 1.3830 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0116 - factorized_top_k/top_10_categorical_accuracy: 0.0255 - factorized_top_k/top_50_categorical_accuracy: 0.1379 - factorized_top_k/top_100_categorical_accuracy: 0.2569 - loss: 71031.6964 - regularization_loss: 0.0000e+00 - total_loss: 71031.6964

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.3719 - factorized_top_k/top_1_categorical_accuracy: 0.0013 - factorized_top_k/top_5_categorical_accuracy: 0.0117 - factorized_top_k/top_10_categorical_accuracy: 0.0264 - factorized_top_k/top_50_categorical_accuracy: 0.1401 - factorized_top_k/top_100_categorical_accuracy: 0.2591 - loss: 70955.8896 - regularization_loss: 0.0000e+00 - total_loss: 70955.8896

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 1.3591 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0119 - factorized_top_k/top_10_categorical_accuracy: 0.0270 - factorized_top_k/top_50_categorical_accuracy: 0.1419 - factorized_top_k/top_100_categorical_accuracy: 0.2617 - loss: 70881.9184 - regularization_loss: 0.0000e+00 - total_loss: 70881.9184

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 1.3647 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0120 - factorized_top_k/top_10_categorical_accuracy: 0.0275 - factorized_top_k/top_50_categorical_accuracy: 0.1438 - factorized_top_k/top_100_categorical_accuracy: 0.2642 - loss: 68996.1852 - regularization_loss: 0.0000e+00 - total_loss: 68996.1852

10/10 [==============================] - 3s 312ms/step - root_mean_squared_error: 1.3647 - factorized_top_k/top_1_categorical_accuracy: 0.0012 - factorized_top_k/top_5_categorical_accuracy: 0.0120 - factorized_top_k/top_10_categorical_accuracy: 0.0275 - factorized_top_k/top_50_categorical_accuracy: 0.1438 - factorized_top_k/top_100_categorical_accuracy: 0.2642 - loss: 67453.3125 - regularization_loss: 0.0000e+00 - total_loss: 67453.3125


Epoch 3/3


 1/10 [==>...........................] - ETA: 2s - root_mean_squared_error: 1.2133 - factorized_top_k/top_1_categorical_accuracy: 6.1035e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0149 - factorized_top_k/top_10_categorical_accuracy: 0.0319 - factorized_top_k/top_50_categorical_accuracy: 0.1605 - factorized_top_k/top_100_categorical_accuracy: 0.2759 - loss: 70038.7500 - regularization_loss: 0.0000e+00 - total_loss: 70038.7500

 2/10 [=====>........................] - ETA: 2s - root_mean_squared_error: 1.2771 - factorized_top_k/top_1_categorical_accuracy: 0.0015 - factorized_top_k/top_5_categorical_accuracy: 0.0177 - factorized_top_k/top_10_categorical_accuracy: 0.0347 - factorized_top_k/top_50_categorical_accuracy: 0.1646 - factorized_top_k/top_100_categorical_accuracy: 0.2874 - loss: 69957.9883 - regularization_loss: 0.0000e+00 - total_loss: 69957.9883    

 3/10 [========>.....................] - ETA: 2s - root_mean_squared_error: 1.2482 - factorized_top_k/top_1_categorical_accuracy: 0.0015 - factorized_top_k/top_5_categorical_accuracy: 0.0179 - factorized_top_k/top_10_categorical_accuracy: 0.0356 - factorized_top_k/top_50_categorical_accuracy: 0.1689 - factorized_top_k/top_100_categorical_accuracy: 0.2915 - loss: 69894.2109 - regularization_loss: 0.0000e+00 - total_loss: 69894.2109

 4/10 [===========>..................] - ETA: 1s - root_mean_squared_error: 1.2471 - factorized_top_k/top_1_categorical_accuracy: 0.0016 - factorized_top_k/top_5_categorical_accuracy: 0.0184 - factorized_top_k/top_10_categorical_accuracy: 0.0374 - factorized_top_k/top_50_categorical_accuracy: 0.1713 - factorized_top_k/top_100_categorical_accuracy: 0.2958 - loss: 69838.3496 - regularization_loss: 0.0000e+00 - total_loss: 69838.3496

 5/10 [==============>...............] - ETA: 1s - root_mean_squared_error: 1.2274 - factorized_top_k/top_1_categorical_accuracy: 0.0017 - factorized_top_k/top_5_categorical_accuracy: 0.0181 - factorized_top_k/top_10_categorical_accuracy: 0.0373 - factorized_top_k/top_50_categorical_accuracy: 0.1727 - factorized_top_k/top_100_categorical_accuracy: 0.2990 - loss: 69772.4859 - regularization_loss: 0.0000e+00 - total_loss: 69772.4859

 6/10 [=================>............] - ETA: 1s - root_mean_squared_error: 1.2212 - factorized_top_k/top_1_categorical_accuracy: 0.0019 - factorized_top_k/top_5_categorical_accuracy: 0.0187 - factorized_top_k/top_10_categorical_accuracy: 0.0384 - factorized_top_k/top_50_categorical_accuracy: 0.1743 - factorized_top_k/top_100_categorical_accuracy: 0.3015 - loss: 69716.7031 - regularization_loss: 0.0000e+00 - total_loss: 69716.7031

 7/10 [====================>.........] - ETA: 0s - root_mean_squared_error: 1.2083 - factorized_top_k/top_1_categorical_accuracy: 0.0018 - factorized_top_k/top_5_categorical_accuracy: 0.0190 - factorized_top_k/top_10_categorical_accuracy: 0.0387 - factorized_top_k/top_50_categorical_accuracy: 0.1753 - factorized_top_k/top_100_categorical_accuracy: 0.3010 - loss: 69678.1395 - regularization_loss: 0.0000e+00 - total_loss: 69678.1395

 8/10 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.2027 - factorized_top_k/top_1_categorical_accuracy: 0.0017 - factorized_top_k/top_5_categorical_accuracy: 0.0190 - factorized_top_k/top_10_categorical_accuracy: 0.0391 - factorized_top_k/top_50_categorical_accuracy: 0.1755 - factorized_top_k/top_100_categorical_accuracy: 0.3018 - loss: 69648.9043 - regularization_loss: 0.0000e+00 - total_loss: 69648.9043

 9/10 [==========================>...] - ETA: 0s - root_mean_squared_error: 1.1947 - factorized_top_k/top_1_categorical_accuracy: 0.0016 - factorized_top_k/top_5_categorical_accuracy: 0.0191 - factorized_top_k/top_10_categorical_accuracy: 0.0392 - factorized_top_k/top_50_categorical_accuracy: 0.1761 - factorized_top_k/top_100_categorical_accuracy: 0.3026 - loss: 69617.2075 - regularization_loss: 0.0000e+00 - total_loss: 69617.2075

10/10 [==============================] - ETA: 0s - root_mean_squared_error: 1.1934 - factorized_top_k/top_1_categorical_accuracy: 0.0016 - factorized_top_k/top_5_categorical_accuracy: 0.0190 - factorized_top_k/top_10_categorical_accuracy: 0.0394 - factorized_top_k/top_50_categorical_accuracy: 0.1771 - factorized_top_k/top_100_categorical_accuracy: 0.3037 - loss: 67792.2855 - regularization_loss: 0.0000e+00 - total_loss: 67792.2855

10/10 [==============================] - 3s 309ms/step - root_mean_squared_error: 1.1934 - factorized_top_k/top_1_categorical_accuracy: 0.0016 - factorized_top_k/top_5_categorical_accuracy: 0.0190 - factorized_top_k/top_10_categorical_accuracy: 0.0394 - factorized_top_k/top_50_categorical_accuracy: 0.1771 - factorized_top_k/top_100_categorical_accuracy: 0.3037 - loss: 66299.1676 - regularization_loss: 0.0000e+00 - total_loss: 66299.1676


1/5 [=====>........................] - ETA: 2s - root_mean_squared_error: 1.1232 - factorized_top_k/top_1_categorical_accuracy: 7.3242e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0076 - factorized_top_k/top_10_categorical_accuracy: 0.0210 - factorized_top_k/top_50_categorical_accuracy: 0.1272 - factorized_top_k/top_100_categorical_accuracy: 0.2361 - loss: 32467.7871 - regularization_loss: 0.0000e+00 - total_loss: 32467.7871

2/5 [===========>..................] - ETA: 0s - root_mean_squared_error: 1.1232 - factorized_top_k/top_1_categorical_accuracy: 6.1035e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0071 - factorized_top_k/top_10_categorical_accuracy: 0.0198 - factorized_top_k/top_50_categorical_accuracy: 0.1224 - factorized_top_k/top_100_categorical_accuracy: 0.2329 - loss: 32508.1611 - regularization_loss: 0.0000e+00 - total_loss: 32508.1611

3/5 [=================>............] - ETA: 0s - root_mean_squared_error: 1.1179 - factorized_top_k/top_1_categorical_accuracy: 5.6966e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0081 - factorized_top_k/top_10_categorical_accuracy: 0.0206 - factorized_top_k/top_50_categorical_accuracy: 0.1224 - factorized_top_k/top_100_categorical_accuracy: 0.2350 - loss: 32508.4043 - regularization_loss: 0.0000e+00 - total_loss: 32508.4043

4/5 [=======================>......] - ETA: 0s - root_mean_squared_error: 1.1130 - factorized_top_k/top_1_categorical_accuracy: 6.7139e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0080 - factorized_top_k/top_10_categorical_accuracy: 0.0204 - factorized_top_k/top_50_categorical_accuracy: 0.1213 - factorized_top_k/top_100_categorical_accuracy: 0.2343 - loss: 32490.4722 - regularization_loss: 0.0000e+00 - total_loss: 32490.4722

5/5 [==============================] - ETA: 0s - root_mean_squared_error: 1.1100 - factorized_top_k/top_1_categorical_accuracy: 9.5000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0086 - factorized_top_k/top_10_categorical_accuracy: 0.0210 - factorized_top_k/top_50_categorical_accuracy: 0.1237 - factorized_top_k/top_100_categorical_accuracy: 0.2349 - loss: 31641.5199 - regularization_loss: 0.0000e+00 - total_loss: 31641.5199

5/5 [==============================] - 1s 190ms/step - root_mean_squared_error: 1.1100 - factorized_top_k/top_1_categorical_accuracy: 9.5000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0086 - factorized_top_k/top_10_categorical_accuracy: 0.0210 - factorized_top_k/top_50_categorical_accuracy: 0.1237 - factorized_top_k/top_100_categorical_accuracy: 0.2349 - loss: 31075.5518 - regularization_loss: 0.0000e+00 - total_loss: 31075.5518


Retrieval top-100 accuracy: 0.235.
Ranking RMSE: 1.110.


The result is a model that performs roughly as well on both tasks as each specialized model.

### Making prediction

We can use the trained multitask model to get trained user and movie embeddings, as well as the predicted rating:

In [15]:
trained_movie_embeddings, trained_user_embeddings, predicted_rating = model({
      "user_id": np.array(["42"]),
      "movie_title": np.array(["Dances with Wolves (1990)"])
  })
print("Predicted rating:")
print(predicted_rating)

Predicted rating:
tf.Tensor([[4.604047]], shape=(1, 1), dtype=float32)


While the results here do not show a clear accuracy benefit from a joint model in this case, multi-task learning is in general an extremely useful tool. We can expect better results when we can transfer knowledge from a data-abundant task (such as clicks) to a closely related data-sparse task (such as purchases).